BUG/ENH - base argument no longer ignored in period resample

ArtinSarraf opened the pull request:
- [ ] closes #23882
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Commit:
ENH - account for base argument in period resample

Commit:
BUG - closer on perfectly emulating previous behavior

Commit:
BUG - all original resample tests now pass

Commit:
TST - add preliminary tests for pandas period resample with base

Commit:
CLN - pep8 adherence

Commit:
TST - add a couple more tests

gfyoung reviewed (changes requested):
Good start!  Also needs a `whatsnew` entry.

Commit:
TST - parameterize tests

ArtinSarraf commented on pandas/tests/resample/test_period_index.py:
```diff
@@ -740,3 +740,33 @@ def test_resample_with_only_nat(self):
+                                                        freq='2Min'))
+        pr = s.resample('5Min', base=3).mean().to_timestamp()
+        tr = s.to_timestamp().resample('5Min', base=3).mean()
+        assert_series_equal(pr, tr)
```
fixed

ArtinSarraf commented on pandas/core/resample.py:
```diff
@@ -1519,9 +1519,24 @@ def _get_period_bins(self, ax):
+            j = -1 if self.freq.onOffset(end) else None
+        else:
+            p_start, p_end = start, end
+            i = j = None
```
if base is non 0: 
I use the same method to get the base adjusted bin edge labels as for a datetimeindex.
Depending on if the start and end dates are on the relevant freq the edges will be extended away from the true start/end labels. 
The i and j indexers are used on the resultant PeriodIndex to adjust them to the proper start/end. 

If base is 0:
then everything is treated the same as before, the start and end date don't change and no slicing of the resultant PeriodIndex is needed so i and j are set to None since `idx[slice(None, None)]` gives the full index

Everything up to here only accounts for what the labels of resultant resample will be, not the actual content being aggregated. For that the bins edges must be adjusted to reflect the base parameter which is done with `rng -= ((freq_mult - self.base) % freq_mult)`

gfyoung replied:
Great!  So now take this explanation and add it as a comment in your code.

Commit:
DOC- add whatsnew entry

gfyoung commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -280,6 +280,7 @@ Other Enhancements
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support a ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times. See :ref:`timeseries.timezone_nonexistent` (:issue:`22647`)
 - :class:`Resampler` now is iterable like :class:`GroupBy` (:issue:`15314`).
 - :meth:`Series.resample` and :meth:`DataFrame.resample` have gained the :meth:`Resampler.quantile` (:issue:`15023`).
+- :meth:`TimeGrouper._get_period_bins` now accounts for the base argument passed through the :class:`PeriodIndexResampler` (:issue:`23882`)
```
We generally don't reference private methods in our `whatsnew` entries.  Is it possible to describe the fix without referencing it?  A good way to come up with a description is look at your test and see what new behavior is fixed based on what you wrote.

gfyoung commented on pandas/core/resample.py:
```diff
@@ -1519,9 +1519,24 @@ def _get_period_bins(self, ax):
 
         start = ax.min().asfreq(self.freq, how=self.convention)
         end = ax.max().asfreq(self.freq, how='end')
+        if self.base:
```
Conversation about that in progress above [here](https://github.com/pandas-dev/pandas/pull/23941#discussion_r236937043).

Commit:
DOC - add comments and modify whatsnew

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -280,6 +280,7 @@ Other Enhancements
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support a ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times. See :ref:`timeseries.timezone_nonexistent` (:issue:`22647`)
 - :class:`Resampler` now is iterable like :class:`GroupBy` (:issue:`15314`).
 - :meth:`Series.resample` and :meth:`DataFrame.resample` have gained the :meth:`Resampler.quantile` (:issue:`15023`).
+- :class:`PeriodIndexResampler` aggregations will now respect the base argument in the same fashion as :class:`DatetimeIndexResampler` (:issue:`23882`)
```
can you put double backticks around base

ArtinSarraf replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1519,9 +1519,27 @@ def _get_period_bins(self, ax):
+            start = start.to_timestamp()
+            end = end.to_timestamp()
+
+            # get base adjusted bin edge labels
```
I would move the PI creation into the if & the else blocks (as the else is almost a trivial creation then).

ArtinSarraf replied:
agreed and done, despite the duplication of the PI creation I think it makes it just that much more obvious/readable.

Commit:
CLN - minor refactor of label creation and whatsnew

jreback commented on pandas/core/resample.py:
```diff
@@ -1519,8 +1519,26 @@ def _get_period_bins(self, ax):
+            j = -1 if self.freq.onOffset(end) else None
+
+            labels = binner = PeriodIndex(start=p_start, end=p_end,
+                                      freq=self.freq,
```
this is a lint error

jreback commented on pandas/core/resample.py:
```diff
@@ -1519,8 +1519,26 @@ def _get_period_bins(self, ax):
+                                              closed=self.closed,
+                                              base=self.base)
+
+            # compensate if edge labels are extened away from true labels
```
can you push this into _get_range_edges w/o breaking the world?

ArtinSarraf replied:
Done. also added some tests to cover the start date not landing onOffset

Commit:
CLN - pep8 adherence

Commit:
Merge branch 'master' into bug/period_resample_base

Commit:
BUG - fix cases where period doesnt start at the 0 base of the freq

Commit:
CLN - remove unneccesary code

Commit:
CLN - pep8 adherence

ArtinSarraf commented:
@jreback - caught a bug in my original bin offset calculation. Previously it was only valid when the index started at the 0 base of the index's frequency. I've fixed that and added several test cases that previously would have failed.

Commit:
Merge branch 'master' into bug/period_resample_base

jreback commented:
can you merge master

Commit:
Merge branch 'master' into bug/period_resample_base

Commit:
BUG - reset resample file

Commit:
BUG - add original changes back in after master reset

Commit:
BUG - add case back in where not start is not onOffset and add tests for this case

Commit:
ENH - allow for edge compensation in _get_range_edges

jreback commented on pandas/core/resample.py:
```diff
@@ -1582,26 +1608,56 @@ def _take_new_index(obj, indexer, new_index, axis=0):
 
 
-def _get_range_edges(first, last, offset, closed='left', base=0):
+def _get_range_edges(first, last, offset, closed='left', base=0,
```
woa, you really need 2 additional kwargs here? there is something very strange if that is the case. Pls try to simplify things.

ArtinSarraf replied:
fixed, no longer needs kwargs. Can take periods now and will infer the adjustment from the input type.

Commit:
BUG/CLN - offsets.Day(n>2) not properly anchoring dates, and make code more readable

ArtinSarraf commented:
@jreback - I also added the fix for this issue since I was already modifying _get_range_edges. Didn't think it was necessary to open a brand new PR, but let me know if it is.

https://github.com/pandas-dev/pandas/issues/24127

Commit:
TST - update test to reflect more up-to-date assumption of what expected output should be

Commit:
Merge branch 'master' into bug/period_resample_base

mroeschke commented on pandas/core/resample.py:
```diff
@@ -1583,25 +1603,46 @@ def _take_new_index(obj, indexer, new_index, axis=0):
 
 
 def _get_range_edges(first, last, offset, closed='left', base=0):
+    if type(first) != type(last):
```
I think its better if you just create a new `_get_period_range_edges` function with the period logic and call the original function `_get_datetime_range_edges`. The function is looking unwieldy.

ArtinSarraf replied:
agreed and fixed.

mroeschke commented:
I would make a new PR for #24127, specially since there is a test that needed changing which warrants discussion. It's better to limit the scope of this PR.

Commit:
CLN - rever changes for other resample fix, to be fixed in another PR

Commit:
CLN - split up range edge functions for timestamp and period

Commit:
CLN - pep8 adherence

ArtinSarraf commented:
@mroeschke - reverted the changes here and added a new PR.
https://github.com/pandas-dev/pandas/pull/24159

Commit:
Merge branch 'master' into bug/period_resample_base

ArtinSarraf commented:
@jreback / @mroeschke - anything else to consider?

Commit:
TST/DOC - add tests for get range edges and update whatsnew

Commit:
Merge branch 'master' into bug/period_resample_base

Commit:
Merge branch 'master' into bug/period_resample_base

Commit:
BUG - account for new period diff behavior

ArtinSarraf commented on pandas/core/resample.py:
```diff
@@ -1606,6 +1631,24 @@ def _get_range_edges(first, last, offset, closed='left', base=0):
 
 
+def _get_period_range_edges(first, last, offset, closed='left', base=0):
+    if not all(isinstance(obj, pd.Period) for obj in [first, last]):
```
done. Not nearly exhaustive, but gets some of the more significant cases

mroeschke replied:
Ah sorry, I meant a case that hits the error message. i.e mixing a Timestamp/Period and other scalar (same with the other case).

ArtinSarraf replied:
done.

jreback replied:
can you add a doc-string

jreback replied:
same as above, need Parmaters / Returns

ArtinSarraf replied:
done.

ArtinSarraf commented on pandas/core/resample.py:
```diff
@@ -1582,7 +1603,11 @@ def _take_new_index(obj, indexer, new_index, axis=0):
 
-def _get_range_edges(first, last, offset, closed='left', base=0):
+def _get_timestamp_range_edges(first, last, offset, closed='left', base=0):
+    if not all(isinstance(obj, pd.Timestamp) for obj in [first, last]):
```
done.

jreback replied:
can you add a doc-string

jreback replied:
can you add Parameters / Returns

ArtinSarraf replied:
done.

ArtinSarraf commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -352,6 +352,7 @@ Other Enhancements
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support a ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times. See :ref:`timeseries.timezone_nonexistent` (:issue:`22647`)
 - :class:`Resampler` now is iterable like :class:`GroupBy` (:issue:`15314`).
 - :meth:`Series.resample` and :meth:`DataFrame.resample` have gained the :meth:`Resampler.quantile` (:issue:`15023`).
+- :class:`PeriodIndexResampler` aggregations will now respect the ``base`` argument in the same fashion as :class:`DatetimeIndexResampler` (:issue:`23882`)
```
fixed

Commit:
TST - add test for bad input to get range edges

Commit:
TST - add one more test case to bad get_range_edges

mroeschke commented on pandas/tests/resample/test_datetime_index.py:
```diff
@@ -1463,3 +1464,37 @@ def f(data, add_arg):
+    ])
+    def test_get_period_range_edges_bad_input(self, first, last):
+        with pytest.raises(TypeError, match='instances of type Timestamp'):
+            _get_timestamp_range_edges(first, last, pd.offsets.Day())
```
Actually, since these functions are only used once and are private, can these hit be simply calling `Series.resample` or `DataFrame.resample`? If they cannot, we may not need to check that first and last are the same scalar type. (same as the other test)

jreback commented on pandas/core/resample.py:
```diff
@@ -1540,20 +1541,39 @@ def _get_period_bins(self, ax):
+                                                   closed=self.closed,
+                                                   base=self.base)
+
+            # Get offset for bin edge (not label edge) adjustment
```
why is this still needed? shouldn't this be in get_period_range_edges?

ArtinSarraf replied:
See the last paragraph of this comment. 
https://github.com/pandas-dev/pandas/pull/23941#discussion_r237332194

Essentially, get_range_edges only accounts for the resultant labels, not the actual bins being aggregated. So the bins edges need to be shifted to reflect the base parameter.

Commit:
DOC - add docstrings

Commit:
TST - remove unneccesary test

jreback commented on pandas/core/resample.py:
```diff
@@ -1582,7 +1602,15 @@ def _take_new_index(obj, indexer, new_index, axis=0):
+def _get_timestamp_range_edges(first, last, offset, closed='left', base=0):
+    """
+    Adjust the provided Timestamp range edge values to the appropriate edge
+    values for the given offset parameters.
```
can you add a Parameters / Returns section here as well (and on _get_period_range_edges)

ArtinSarraf replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1606,6 +1634,28 @@ def _get_range_edges(first, last, offset, closed='left', base=0):
 
+def _get_period_range_edges(first, last, offset, closed='left', base=0):
+    """
+    Adjust the provided Period range edge values to the appropriate edge
```
see comment above

ArtinSarraf replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1606,6 +1634,28 @@ def _get_range_edges(first, last, offset, closed='left', base=0):
+def _get_period_range_edges(first, last, offset, closed='left', base=0):
+    """
+    Adjust the provided Period range edge values to the appropriate edge
+    values for the given offset parameters.
```
can you give a 1 sentence on what this is doing. imagine you have never seen this and explain in simplest possible terms.

ArtinSarraf replied:
updated, please take a look and provide any feedback. the timestamp one is a couple sentences, but I think it's clear.

jreback reviewed (changes requested):
looks good. just want some more docs.

Commit:
DOC - update docstrings

Commit:
Merge branch 'master' into bug/period_resample_base

Commit:
CLN - rename offset to bin_shift, so as not to confuse with pd.offsets

Commit:
DOC - add params and returns sections

Commit:
Merge branch 'master' into bug/period_resample_base

jreback commented:
thanks @ArtinSarraf
